File: src/test/java/org/cloudfoundry/promregator/config/validations/CompatibleCAPIVersionTest.java
package org.cloudfoundry.promregator.config.validations;

import org.cloudfoundry.promregator.cfaccessor.CFAccessor;
import org.cloudfoundry.promregator.config.PromregatorConfiguration;
import org.cloudfoundry.promregator.config.Target;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.LinkedList;
import java.util.List;

import static org.mockito.Mockito.when;

class CompatibleCAPIVersionTest {
	@InjectMocks
	private CompatibleCAPIVersion subject;

	@Mock
	private CFAccessor cfAccessorMock;

	@BeforeEach
	public void initMocks(){
		MockitoAnnotations.openMocks(this);
	}

	@Test
	void testValidateConfigNoV3() {
		Target t = new Target();
		t.setKubernetesAnnotations(true);

		List<Target> targets = new LinkedList<>();
		targets.add(t);

		PromregatorConfiguration pc = new PromregatorConfiguration();
		pc.setTargets(targets);

		when(this.cfAccessorMock.isV3Enabled()).thenReturn(false);
		String result = subject.validate(pc);

		Assertions.assertNotNull(result);
	}

	@Test
	void testValidateConfigNoV3NoKubernetes() {
		Target t = new Target();
		t.setKubernetesAnnotations(false);

		List<Target> targets = new LinkedList<>();
		targets.add(t);

		PromregatorConfiguration pc = new PromregatorConfiguration();
		pc.setTargets(targets);

		when(this.cfAccessorMock.isV3Enabled()).thenReturn(false);
		String result = subject.validate(pc);

		Assertions.assertNull(result);
	}

	@Test
	void testValidateConfigOk() {
		Target t = new Target();
		t.setKubernetesAnnotations(true);

		List<Target> targets = new LinkedList<>();
		targets.add(t);

		PromregatorConfiguration pc = new PromregatorConfiguration();
		pc.setTargets(targets);

		when(this.cfAccessorMock.isV3Enabled()).thenReturn(true);
		String result = subject.validate(pc);

		Assertions.assertNull(result);
	}
}

File: src/test/java/org/cloudfoundry/promregator/config/validations/CompatibleCAPIVersionTest.kt
package org.cloudfoundry.promregator.config.validations

import io.mockk.every
import io.mockk.mockk
import org.cloudfoundry.promregator.cfaccessor.CFAccessor
import org.cloudfoundry.promregator.lite.config.PromregatorConfiguration
import org.cloudfoundry.promregator.lite.config.CfTarget
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class CompatibleCAPIVersionTest {
    lateinit var sut: CompatibleCAPIVersion

    lateinit var cfAccessorMock: CFAccessor
    @BeforeEach
    fun initMocks() {
        cfAccessorMock = mockk()
        sut = CompatibleCAPIVersion(cfAccessorMock)
    }

    @Test
    fun testValidateConfigNoV3() {
        val t = CfTarget(kubernetesAnnotations = true)
        val pc = PromregatorConfiguration(listOf(t))

        every { cfAccessorMock.isV3Enabled } returns false
        val result = sut.validate(pc)
        Assertions.assertNotNull(result)
    }

    @Test
    fun testValidateConfigNoV3NoKubernetes() {
        val t = CfTarget(kubernetesAnnotations = false)
        val pc = PromregatorConfiguration(listOf(t))
        every { cfAccessorMock.isV3Enabled } returns false
        val result = sut.validate(pc)
        Assertions.assertNull(result)
    }

    @Test
    fun testValidateConfigOk() {
        val t = CfTarget(kubernetesAnnotations = true)
        val pc = PromregatorConfiguration(listOf(t))
        every { cfAccessorMock.isV3Enabled } returns true
        val result = sut.validate(pc)
        Assertions.assertNull(result)
    }
}
